frontend/src/metabase/selectors/metadata.js

```diff
@@ -123,8 +124,36 @@ export const getSegments = createSelector(
 
 // MISC
 
-export const getParameterFieldValues = (state, props) => {
-    return getFieldValues(getIn(state, ["metadata", "fields", props.parameter.field_id]));
+// Returns a dictionary of field id:s mapped to matching field values
+const getParameterFieldValuesByFieldId = (state, props) => _.chain(getFields(state))
+    .pick(getFields(state), ...props.parameter.field_ids)
+    .mapObject(getFieldValues)
+    .value()
+
+// Check if the lengths of field value arrays equal for each field
+// TODO: Why can't we use plain shallowEqual, i.e. why the field value arrays change very often?
+const createFieldValuesEqualSelector = createSelectorCreator(defaultMemoize, (a, b) => {
+    return shallowEqual(_.mapObject(a, (values) => values.length), _.mapObject(b, (values) => values.length));
+})
+
+// Merges the field values of fields linked to a parameter and removes duplicates
+// TODO Atte Keinänen 7/20/17: How should this work for remapped values?
+// TODO Atte Keinänen 7/20/17: Should we have any thresholds if the count of field values is high or we have many (>2?) fields?
+export const makeGetMergedParameterFieldValues = () => {
+    return createFieldValuesEqualSelector(getParameterFieldValuesByFieldId, (fieldValues) => {
```

Comment:
why is this a function creator rather than just a function?

Reply:
If we just have a single function call of `createSelector` (or in this case the customized version called `createFieldValuesEqualSelector`) and export it, then we create only one selector that is reused everywhere in the app. 

This works well if all React components in the app call the selector with same parameters. In our case each instance of `<ParameterValueWidget>` expects different list of field values, and that list depends on `field_ids` array which `<ParameterValueWidget>` receives via React props.

We want that each `<ParameterValueWidget>` has its own selector because otherwise the parameters of one global selector change all the time and its never able to properly memoize anything. That leads to a terrible UI performance (I mean, it really is terrible, I tested it).

I should probably add an explanatory comment to the code above, in minimum by adding a link to https://github.com/reactjs/reselect#sharing-selectors-with-props-across-multiple-components and explaining that `fields_ids` value is specific to each ParameterValueWidget instance.

Reply:
And some clarifications:

- The old approach was fast because it basically just used the field values list that the backend returned as-is. In that case any `reselect` stuff was useless.

- The new approach does a relatively heavyweight merge by sorting and then removing duplicates of combined field value lists (one list per each field id). That's why without `reselect` and the equality comparison / selector creator nastiness the UI performance becomes terrible.